src/simulation.hpp

```diff
@@ -2950,7 +2948,46 @@ template <typename problem_t> void AMRSimulation<problem_t>::doDiagnostics()
 
 	for (const auto &diag : m_diagnostics) {
 		if (diag->doDiag(tNew_[0], istep[0])) {
-			diag->processDiag(istep[0], tNew_[0], GetVecOfConstPtrs(diagMFVec), m_diagVars, simulationMetadata_);
+			// Check if this is a DiagPlotfile - if so, call its special writePlotfile method
+			auto *plotfileDiag = dynamic_cast<DiagPlotfile *>(diag.get());
+			if (plotfileDiag != nullptr) {
+				// Prepare full plotfile data
+				int const included_ghosts = std::min(nghost_cc_, nghost_fc_);
+				amrex::Vector<amrex::MultiFab> mf_cc = PlotFileMF_cc(included_ghosts);
+				amrex::Vector<const amrex::MultiFab *> mf_cc_ptr = amrex::GetVecOfConstPtrs(mf_cc);
+				auto const varnames = GetPlotfileVarNames();
+
+				// Prepare face-centered data
+				std::array<amrex::Vector<amrex::MultiFab>, AMREX_SPACEDIM> mf_fc = PlotFileMF_fc(nghost_fc_);
+				std::array<amrex::Vector<const amrex::MultiFab *>, AMREX_SPACEDIM> mf_fc_ptr;
+				for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+					mf_fc_ptr[idim] = amrex::GetVecOfConstPtrs(mf_fc[idim]);
+				}
+				auto const varnames_fc = GetPlotfileVarNames_fc();
+
+				// Write cell-centered data and particles
+				plotfileDiag->writePlotfile(istep[0], tNew_[0], finestLevel(), mf_cc_ptr, varnames, Geom(0, finestLevel()), istep, refRatio(),
+							    particleRegister_, do_tracers, TracerPC.get(), simulationMetadata_);
+
+				// Write face-centered data if present
+				plotfileDiag->writePlotfileFC<problem_t>(amrex::Concatenate(plotfileDiag->getDiagFileName(), istep[0], 5), finestLevel(),
+									 mf_fc_ptr, varnames_fc, Geom(0, finestLevel()), tNew_[0], istep, refRatio(),
+									 simulationMetadata_);
+			} else {
+				// Check if this is a DiagProjectionPlot - if so, call its special writeProjection method
```

Comment:
Is there any way we can avoid this for the DiagProjectionPlot? It would be better to do this in the processDiag() function.